src/extra_writer/writer.py

```
@@ -0,0 +1,432 @@
+            if new_class._meta.class_attrs_interface:
+                setattr(new_class, ds_name, ds.get_attribute_setter(ds_name))
+            else:
+                setattr(new_class, ds_name, BlockedSetter())
+
+        return new_class
+
+
+class FileWriterBase(object):
+    """Writes data in EuXFEL format"""
+    datasets = {}
+
+    def __new__(cls, *args, **kwargs):
+        if not cls.datasets:
+            raise TypeError(f"Can't instantiate class {cls.__name__}, "
+                            "because it has no datasets")
+        return super().__new__(cls)
+
+    def __init__(self, filename, **kwargs):
+        self._ds_attrs = {}
+        self._train_data = {}
+        self.trains = []
+        self.timestamp = []
+        self.flags = []
+        self.seq = 0
+        self.filename = filename
+        self.param = kwargs
+
+        source_decl = {}
+        for ds_name, ds in self.datasets.items():
+            ds.init_dataset_attr(self)
+            ds_attr = ds(self)
+            repr_ds = source_decl.setdefault(ds_attr.source_name, ds)
+            if issubclass(ds.source_class, repr_ds.source_class):
+                source_decl[ds_attr.source_name] = ds
+            elif not issubclass(repr_ds.source_class, ds.source_class):
+                raise TypeError("Incompatible datasets in the source "
+                                f"'{ds_attr.source_name}'")
+
+        self.nsource = len(source_decl)
+        self.sources = {}
+        self.list_of_sources = []
+        self.source_ntrain = {}
+        for src_name, ds in source_decl.items():
+            src = ds.source(self)
+            self.sources[src_name] = src
+            self.source_ntrain[src_name] = 0
+            self.list_of_sources.append((src.section, src_name))
+
+        for dsname, ds in self.datasets.items():
+            src_name = ds(self).source_name
+            self.sources[src_name].add(dsname, ds)
+
+        file = h5py.File(filename.format(seq=self.seq), 'w')
```

If a `{seq}` field is not in the filename then this will fail silently. I guess in some cases we won't need multiple files, so maybe if `break_into_sequence` is set then this should throw an exception? Alternatively, we could switch to specifying a prefix instead of a full filename.

---

I've added the check, thanks!